sia/src/rhp/rpc.rs

```diff
@@ -0,0 +1,613 @@
+use crate::encoding::{SiaDecodable, SiaDecode, SiaEncodable, SiaEncode};
+use blake2b_simd::Params;
+use serde::{Deserialize, Serialize};
+use time::OffsetDateTime;
+
+use crate::signing::{PrivateKey, PublicKey, Signature};
+use crate::types::v2::{FileContract, SatisfiedPolicy, SiacoinElement, SiacoinInput, Transaction};
+use crate::types::{Address, ChainIndex, Currency, FileContractID, Hash256, Leaf};
+
+/// Contains the prices and parameters of a host.
+#[derive(Debug, PartialEq, Serialize, Deserialize, SiaEncode, SiaDecode)]
+#[serde(rename_all = "camelCase")]
+pub struct HostPrices {
+    /// The price of forming a new contract with the host.
+    pub contract_price: Currency,
+    /// The collateral per byte per block the host will
+    /// risk for stored data.
+    pub collateral: Currency,
+    /// The cost of storing a sector on the host per byte per block.
+    pub storage_price: Currency,
+    /// The cost of uploading data to the host per byte.
+    pub ingress_price: Currency,
+    /// The cost of downloading data from the host per byte.
+    pub egress_price: Currency,
+    /// The cost to remove a sector from a contract.
+    pub free_sector_price: Currency,
+    /// The current height of the host's blockchain.
+    pub tip_height: u64,
+    /// The time until which the prices are valid.
+    pub valid_until: OffsetDateTime,
+
+    pub signature: Signature,
+}
+
+impl HostPrices {
+    /// Computes the signature hash for the host prices.
+    pub fn sig_hash(&self) -> Hash256 {
+        let mut state = Params::new().hash_length(32).to_state();
+        self.contract_price.encode(&mut state).unwrap();
+        self.collateral.encode(&mut state).unwrap();
+        self.storage_price.encode(&mut state).unwrap();
+        self.ingress_price.encode(&mut state).unwrap();
+        self.egress_price.encode(&mut state).unwrap();
+        self.free_sector_price.encode(&mut state).unwrap();
+        self.tip_height.encode(&mut state).unwrap();
+        self.valid_until.encode(&mut state).unwrap();
+        state.finalize().into()
+    }
+
+    /// Checks if the prices are valid for the given host key and timestamp.
+    pub fn is_valid(&self, host_key: &PublicKey, timestamp: OffsetDateTime) -> bool {
+        self.valid_until > timestamp
+            && self.tip_height > 0
+            && host_key.verify(&self.sig_hash(), &self.signature)
+    }
+}
```

Comment:
Don't need it in this PR but this is missing the helpers to compute the RPC costs.